core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/queue/EventUpload.kt

```diff
@@ -137,23 +125,83 @@ internal class EventUpload(
         }
     }
 
-    private fun cleanup(file: File, filePath: String) {
-        storage.remove(file.path).let {
-            LoggerAnalytics.debug("Removed file: $filePath")
+    private suspend fun uploadEvents(batchPayload: String, filePath: String) {
+        var result: EventUploadResult
+        do {
+            val updatedPayload = JsonSentAtUpdater.updateSentAt(batchPayload)
+            LoggerAnalytics.verbose("Batch Payload: $updatedPayload")
+            result = httpClientFactory.sendData(updatedPayload).toEventUploadResult()
+
+            when (result) {
+                is Success -> {
+                    LoggerAnalytics.debug("Event uploaded successfully. Server response: ${result.response}")
+                    maxAttemptsWithBackoff.reset()
+                    cleanup(filePath)
+                }
+                is RetryAbleError -> {
+                    LoggerAnalytics.debug("EventUpload: Retry able error occurred")
+                    maxAttemptsWithBackoff.delayWithBackoff()
+                }
+                is NonRetryAbleError -> {
+                    maxAttemptsWithBackoff.reset()
+                    handleNonRetryAbleError(result, filePath)
+                }
+            }
+        } while (result is RetryAbleError)
+    }
+
+    @OptIn(UseWithCaution::class)
+    private fun handleNonRetryAbleError(status: NonRetryAbleError, filePath: String) {
+        when (status) {
+            NonRetryAbleEventUploadError.ERROR_400 -> {
+                LoggerAnalytics.error(
+                    "Invalid request: missing or malformed body. " +
+                        "Ensure the payload is valid JSON and includes either anonymousId or userId."
+                )
+                cleanup(filePath)
+            }
+
+            NonRetryAbleEventUploadError.ERROR_401 -> {
+                LoggerAnalytics.error("Invalid write key. Ensure the write key is valid.")
+                cancel()
+                analytics.handleInvalidWriteKey()
+            }
+
+            NonRetryAbleEventUploadError.ERROR_404 -> {
+                LoggerAnalytics.error("Source is disabled. Stopping the upload process until the source is enabled again.")
+                cancel()
+                analytics.sourceConfigState.dispatch(SourceConfig.DisableSourceAction())
+            }
+
+            NonRetryAbleEventUploadError.ERROR_413 -> {
+                LoggerAnalytics.error("Batch request failed: Payload size exceeds the maximum allowed limit.")
```

Comment:
Since you didn't use `batch` anywhere else.
```suggestion
                LoggerAnalytics.error("Request failed: Payload size exceeds the maximum allowed limit.")
```